src/main/java/net/imglib2/parallel/DefaultTaskExecutor.java

```diff
@@ -0,0 +1,183 @@
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+package net.imglib2.parallel;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.Future;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+/**
+ * A {@link TaskExecutor} that wraps around a given {@link ExecutorService}.
+ */
+public class DefaultTaskExecutor implements TaskExecutor
+{
+
+	private final ExecutorService executorService;
+
+	public DefaultTaskExecutor( ExecutorService executorService )
+	{
+		this.executorService = executorService;
+	}
+
+	@Override
+	public ExecutorService getExecutorService()
+	{
+		return executorService;
+	}
+
+	@Override
+	public int getParallelism()
+	{
+		if ( executorService instanceof ForkJoinPool )
+			return ( ( ForkJoinPool ) executorService ).getParallelism();
+		else if ( executorService instanceof ThreadPoolExecutor )
+			return Math.max( 1, ( ( ThreadPoolExecutor ) executorService ).getCorePoolSize() );
+		else if ( executorService instanceof ForkJoinExecutorService )
+			return ( ( ForkJoinExecutorService ) executorService ).getParallelism();
+		else if ( executorService instanceof SequentialExecutorService )
+			return ( ( SequentialExecutorService ) executorService ).getParallelism();
+		return Runtime.getRuntime().availableProcessors();
+	}
+
+	@Override
+	public void runAll( List< Runnable > tasks )
+	{
+		List< Callable< Object > > callables = new ArrayList<>( tasks.size() );
+		// use for-loop because stream with collect(Collectors.toList) is slow.
+		for ( Runnable task : tasks )
+			callables.add( Executors.callable( task ) );
+		invokeAllIgnoreResults( callables );
+	}
+
+	@Override
+	public int suggestNumberOfTasks()
+	{
+		int parallelism = getParallelism();
+		return ( parallelism == 1 ) ? 1 : ( int ) Math.min( ( long ) parallelism * 4L, ( long ) Integer.MAX_VALUE );
```

Comment:
The max size of a Java array (or an `ArrayList`) is a little smaller than `Integer.MAX_VALUE` and the exact size depends on the JVM. For my JVM (`openjdk version "1.8.0_222"`), for example,
```kotlin
val list = ArrayList<Unit?>(Integer.MAX_VALUE)
```
fails with
```
Caused by: java.lang.OutOfMemoryError: Requested array size exceeds VM limit
        at java.util.ArrayList.<init>(ArrayList.java:153)
        at Max_int_array_list.<init>(Unknown Source)
        ... 13 mor
```

The max size that does not fail in my JVM is `Integer.MAX_VALUE - 3`:
```kotlin
val list = ArrayList<Unit?>(Integer.MAX_VALUE - 3)
```

I am not aware of a way to retrieve the VM limit but nonetheless it would be good to limit the size to a little smaller than `Integer.MAX_VALUE`, e.g. `Integer.MAX_VALUE - 5`.

https://bugs.openjdk.java.net/browse/JDK-8059914?focusedCommentId=13585092&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13585092

Reply:
This is also a problem with `Util.getArrayOrCellImgFactory` currently, no? Because we use `Integer.MAX_VALUE` for the cutoff between `ArrayImg` and `CellImg`.

FWIW, on my macOS system, the maximum array size seems to be `Integer.MAX_VALUE - 2` consistently. Tested empirically with the following AdoptOpenJDK versions:  12, 11.0.2, 10.0.2, 9, 1.8.0_212.

Reply:
Yes, that would be a problem there, too.

Reply:
Yes, we should change that. Actually I also had that on my TODO list already...

In Mastodon, we use [this:](https://github.com/bigdataviewer/mastodon-graph/blob/f432bf7e8a5784e1e8ed079672a060599e78fe63/src/main/java/org/mastodon/undo/UndoDataStack.java#L113-L118)
```java
	/**
	 * The maximum size of array to allocate. Some VMs reserve some header
	 * words in an array. Attempts to allocate larger arrays may result in
	 * OutOfMemoryError: Requested array size exceeds VM limit
	 */
	private static final int MAX_ARRAY_SIZE = Integer.MAX_VALUE - 8;
```
We could put it into a `net.imglib2.Constants` class and then use everywhere where we use `Integer.MAX_VALUE` at the moment. But as @ctrueden suggested, maybe we should make it a method instead of a field, so that it will not be inlined as a compile time constant.

Reply:
A constant might be OK as long as it is `private`. Because the only things that can use it then will be in the same component, and will also be recompiled if it ever changes.

Reply:
I wouldn't change this for now. The purpose of `suggestNumberOfTasks` is to return a reasonable default value for parallelization. Something went wrong, if a value close to `Integer.MAX_VALUE` is returned. And I think an "Requested array size exceeds VM limit" exception will quickly lead to the correct conclusion: `suggestNumberOfTasks` and `getParallism` returned wrong values, that are way too big. The purpose of the line in the code is to prevent integer overflow, which will turn a number that is way to big into a negative number.

There should be a separate issue, that targets the problem that `Integer.MAX_VALUE` is not the upper bound for arrays.